### DISCUSSION TITLE
Recommended regular expression for PurgeCSS and what is NODE_ENV?

### DISCUSSION THREAD

**Discussion by tian-yan** [tailwindlabs/tailwindcss#4592] [Help]
I came across 3 answers in regards to regular expression for PurgeCSS. **Q1: Which one is recommended please ?** I tend to follow the one from official document but just want to double confirm. 
- `defaultExtractor: content => content.match(/[A-Za-z0-9-_:/]+/g)  || []` From [Adam Wathan's Video Tutorial, at 4 mins 37secs](https://www.youtube.com/watch?v=bhoDwo24K5Q&list=PL7CcGwsqRpSM3w9BT_21tUU8JN2SnyckR&index=8)
- `/[^<>"'`\s]*[^<>"'`\s:]/g` from [TailwindCSS official document](https://tailwindcss.com/docs/optimizing-for-production#writing-purgeable-html)
- another answer was [discussed here](https://github.com/tailwindlabs/tailwindcss/discussions/1656#discussioncomment-7876). 

Q2: **NODE_ENV** is mentioned in [official document](https://tailwindcss.com/docs/optimizing-for-production#basic-usage). I am currently using VS Code, how can I tell what **mode** I am in when compiling CSS file ?

**Comment by VicGUTT** [comment]
Hey !

#### About the regular expression
The one you need may depend on your use case, your templating language, if you're using a framework/tool and which one etc...
The video you linked seem to be pre-Tailwind with built in PurgeCSS support. Back then you had to configure PurgeCSS yourself, that is no longer the case nowadays.
So instead you have a particular need, I would suggest not messing with the defaults. I have never needed to do that myself.

#### Your environnement
The environnement you are in and how it's named depends on how your build system configured it. Generally, without even thinking about it, when you are coding locally you are in **development**, when you run a script like `npm run build` or `npm run production` then it will set the environnement to **production**.

You can check your environnement by doing something like `console.log(process.env.NODE_ENV)`.
Also check you `package.json`'s script section for any hints.

Hope it helps :v:

**Reply by tian-yan** [reply]
> Nowadays, I do not have to configure PurgeCSS myself

It only means that I do not have to configure the regular expression manually anymore, and it is already built-in TailwindCSS. But I still have to:
1. install PurgeCSS via `npm install @fullhuman/postcss-purgecss`
2. Add below in `postcss.config.js` file
```
module.exports = {
    plugins: [
        require(`@fullhuman/postcss-purgecss`),
        ]
}
```
3. Add related HTML files in `tailwind.config.js` according to above mentioned official document;
```
module.exports = {
    purge: ['../*/TW_Index.html'],
}
```
**Q1: Is my understanding correct ?**
Above is my current setup for PurgeCSS, but when I run `npm run build` to re-compile the CSS file, below errors pop up.
```
(base) YANs-MBP:src yantian$ npm run build

> src@1.0.0 build
> postcss source.css -o ../publish/tw_style.css

**Error: PurgeCSS plugin does not have the correct options**
    at c (/Users/yantian/Documents/TailwindCSS/src/node_modules/@fullhuman/postcss-purgecss/lib/postcss-purgecss.js:1:1103)
    at /Users/yantian/Documents/TailwindCSS/src/node_modules/postcss-load-config/src/plugins.js:64:18
    at Array.forEach (<anonymous>)
    at plugins (/Users/yantian/Documents/TailwindCSS/src/node_modules/postcss-load-config/src/plugins.js:58:13)
    at processResult (/Users/yantian/Documents/TailwindCSS/src/node_modules/postcss-load-config/src/index.js:33:14)
    at /Users/yantian/Documents/TailwindCSS/src/node_modules/postcss-load-config/src/index.js:94:14
    at async Promise.all (index 0)
(base) YANs-MBP:src yantian$ 

(base) YANs-MBP:src yantian$ npm -v tailwindcss
7.13.0
```

When I switch back to my original setup, in other words, follow the video tutorial mentioned above, it worked perfectly when i re-compile CSS via `npm run build`, the resulted CSS file size was shrunk from 4MB to 13KB. 
```
//postcss.config.js
module.exports = {
    plugins: [
        require('tailwindcss'),
        require('autoprefixer'),
        require(`@fullhuman/postcss-purgecss`)({
            content: [
                `../publish/TW_Index.html`,
            ],
            defaultExtractor: content => content.match(/[A-Za-z0-9-_:/]+/g)  || []
        }),
    ]
}
```
Q2: I am a bit confused on above conflicting testing results.
